kernel/nativelib.ml

```diff
@@ -183,11 +183,24 @@ let call_linker ?(fatal=true) ~prefix f upds =
      else if !Flags.debug then Feedback.msg_debug CErrors.(iprint exn));
   match upds with Some upds -> update_locations upds | _ -> ()
 
-let link_library ~prefix ~dirname ~basename =
+let link_library dirname libname =
+  let prefix = mod_uid_of_dirpath libname ^ "." in
+  let basename = Dynlink.adapt_filename (prefix ^ "cmo") in
   (* We try both [output_dir] and [.coq-native], unfortunately from
      [Require] we don't know if we are loading a library in the build
      dir or in the installed layout *)
   let install_location = dirname / dft_output_dir / basename in
   let build_location = dirname / !output_dir / basename in
   let f = if Sys.file_exists build_location then build_location else install_location in
   call_linker ~fatal:false ~prefix f None
```

Comment:
Isn't this going to wreak havoc w.r.t. dependency computation? The code of `call_linker` registers imperatively the name of the library inside a global table to check that it is available (see `Nativecode.is_code_loaded`). By delaying this registering can't we get into a situation where the dependency check believes that the library is not registered and will recompile it (as per ondemand) and we will get double compilation in the end?

Reply:
I had straced Coq to make sure the correct files were loaded at the correct time, but perhaps I have missed some specific call paths. Let me check.

Reply:
Good catch. The very first call to `native_compute` was indeed performed as if in `ondemand` mode.